python/pyarrow/fs.py

```diff
@@ -63,6 +63,31 @@ def __getattr__(name):
     )
 
 
+def _ensure_filesystem(filesystem, use_mmap=False):
+    if isinstance(filesystem, FileSystem):
+        return filesystem
+
+    # handle fsspec-compatible filesystems
+    try:
+        import fsspec
+        if isinstance(filesystem, fsspec.AbstractFileSystem):
```

Comment:
@martindurant is this the best way to check if a given object is a fsspec filesytem? 

(I suppose that in theory "fsspec-compatible" doesn't mean it needs to sublcass from fsspec, since it's the spec (the methods) that needs to be implemented. But I don't think we can detect "duck-typed" filesystems)

Reply:
Agree completely. So far, all implementations II know about are subclasses, but it's not required. I have not yet got around to implementing an entrypoints way to declare implementations. 
So, you could try to look for some specific methods, but I'm not sure what we could guarantee was specific to fsspec.